projects/Funkwhale/services/funkwhale/module.nix

```diff
@@ -0,0 +1,538 @@
+            else
+              toString value;
+        in
+        attrs:
+        lib.mapAttrs (
+          _: value: if lib.isList value then lib.concatMapStringsSep "," toStr value else toStr value
+        ) attrs;
+      default = { };
+      example = {
+        FUNKWHALE_HOSTNAME = "localhost";
+      };
+      description = ''
+        See <https://docs.funkwhale.audio/administrator/configuration/env-file.html>.
+
+        Secret settings such as `DJANGO_SECRET_KEY` should go in `services.funkwhale.secretEnvironmentFile`.
+
+        You will need to manually create your first user account.
+        Example: `sudo -u funkwhale funkwhale-manage fw users create --superuser`.
+        See <https://docs.funkwhale.audio/administrator/manage-script/users.html> for
+        more information.
+      '';
+    };
+
+    secretEnvironmentFile = lib.mkOption {
+      type = with lib.types; nullOr path;
+      default = null;
+      example = "/run/secrets/funkwhale.env";
+      description = ''
+        Secret environment variable settings. These overwrite `services.funkwhale.settings`.
+        See <https://docs.funkwhale.audio/administrator/configuration/env-file.html>.
+
+        Example:
+
+        ```
+        DJANGO_SECRET_KEY="some-very-increadibly-secret-key"
+        TYPESENSE_API_KEY="some-other-very-increadibly-super-secret-key"
+        ```
+
+        `DJANGO_SECRET_KEY` will be auto-generated if not set here.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !(cfg.settings ? DJANGO_SECRET_KEY);
+        message = ''
+          Setting `services.funkwhale.settings.DJANGO_SECRET_KEY` is not secure.
+          Instead, `DJANGO_SECRET_KEY` should be set in `services.funkwhale.secretEnvironmentFile`.
+        '';
+      }
+    ];
+
+    services.funkwhale.settings = lib.mkMerge [
+      {
+        MEDIA_ROOT = "/var/lib/funkwhale/data/media";
+        MUSIC_DIRECTORY_PATH = "/var/lib/funkwhale/data/music";
+      }
+      (lib.mkIf cfg.createDatabaseLocally {
+        DATABASE_HOST = "";
+        DATABASE_PORT = 5432;
+        DATABASE_USER = cfg.user;
+        # Peer authentication.
+        DATABASE_PASSWORD = "";
+      })
+      (lib.mkIf cfg.createRedisLocally {
+        CACHE_URL = "unix://${config.services.redis.servers.funkwhale.unixSocket}?db=0";
+        CELERY_BROKER_URL = "redis+socket://${config.services.redis.servers.funkwhale.unixSocket}?virtual_host=0";
+      })
+      (lib.mkIf cfg.configureNginx {
+        REVERSE_PROXY_TYPE = "nginx";
+        NGINX_MAX_BODY_SIZE = lib.mkDefault "100M";
+      })
+    ];
+
+    users.users = lib.mkIf (cfg.user == "funkwhale") {
+      ${cfg.user} = {
+        isSystemUser = true;
+        group = cfg.group;
+      };
+    };
+    users.groups = lib.mkIf (cfg.group == "funkwhale") {
+      ${cfg.group} = { };
+    };
+
+    environment.systemPackages = [
+      (pkgs.writeShellScriptBin "funkwhale-manage" ''
```

Comment:
I would move this script creation to let binding to make configuration of `environment.systemPackages` cleaner